store/auth.go

```diff
@@ -26,218 +26,157 @@ import (
 	"io"
 	"io/ioutil"
 	"net/http"
-
-	"gopkg.in/macaroon.v1"
+	"net/url"
+	"sync"
 
 	"github.com/snapcore/snapd/httputil"
+	"github.com/snapcore/snapd/overlord/auth"
 	"github.com/snapcore/snapd/snapdenv"
 )
 
-var (
-	developerAPIBase = storeDeveloperURL()
-	// macaroonACLAPI points to Developer API endpoint to get an ACL macaroon
-	MacaroonACLAPI   = developerAPIBase + "dev/api/acl/"
-	ubuntuoneAPIBase = authURL()
-	// UbuntuoneLocation is the Ubuntuone location as defined in the store macaroon
-	UbuntuoneLocation = authLocation()
-	// UbuntuoneDischargeAPI points to SSO endpoint to discharge a macaroon
-	UbuntuoneDischargeAPI = ubuntuoneAPIBase + "/tokens/discharge"
-	// UbuntuoneRefreshDischargeAPI points to SSO endpoint to refresh a discharge macaroon
-	UbuntuoneRefreshDischargeAPI = ubuntuoneAPIBase + "/tokens/refresh"
-)
-
-// a stringList is something that can be deserialized from a JSON
-// []string or a string, like the values of the "extra" documents in
-// error responses
-type stringList []string
+// An Authorizer can authorize a request using credentials directly or indirectly available.
+type Authorizer interface {
+	// Authorize authorizes the given request.
+	// If implementing multiple kind of authorization at the same
+	// time all they should be performed separately ignoring
+	// errors, as the higher-level code might as well treat Authorize
+	// as best-effort and only log any returned error.
+	Authorize(r *http.Request, dauthCtx DeviceAndAuthContext, user *auth.UserState, opts *AuthorizeOptions) error
+}
 
-func (sish *stringList) UnmarshalJSON(bs []byte) error {
-	var ss []string
-	e1 := json.Unmarshal(bs, &ss)
-	if e1 == nil {
-		*sish = stringList(ss)
-		return nil
-	}
+type AuthorizeOptions struct {
+	// deviceAuth is set if device authorization should be
+	// provided if available.
+	deviceAuth bool
 
-	var s string
-	e2 := json.Unmarshal(bs, &s)
-	if e2 == nil {
-		*sish = stringList([]string{s})
-		return nil
-	}
-
-	return e1
+	apiLevel apiLevel
 }
 
-type ssoMsg struct {
-	Code    string                `json:"code"`
-	Message string                `json:"message"`
-	Extra   map[string]stringList `json:"extra"`
-}
+type RefreshingAuthorizer interface {
+	Authorizer
 
-// returns true if the http status code is in the "success" range (2xx)
-func httpStatusCodeSuccess(httpStatusCode int) bool {
-	return httpStatusCode/100 == 2
+	// Refresh transient authorization data.
+	RefreshAuth(need AuthRefreshNeed, dauthCtx DeviceAndAuthContext, user *auth.UserState, client *http.Client) error
 }
 
-// returns true if the http status code is in the "client-error" range (4xx)
-func httpStatusCodeClientError(httpStatusCode int) bool {
-	return httpStatusCode/100 == 4
+// AuthRefreshNeed represents which authorization data needs refreshing.
+type AuthRefreshNeed struct {
+	Device bool
+	User   bool
```

Comment:
This can also be unexported along with its fields, it's only used in the store package